### PR TITLE
[IMP] [base, account_facturx]: Add PDF/A(-3B) support

### PR DESCRIPTION
**Pull request by vin-odoo**
Improve the factur-x export in two ways: make the exported PDF
PDF/A-3B compliant, and add the factur-x XMP metadata inside the file.

Task id # 2668919

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account_facturx/data/facturx_templates.xml
Comment (smetl) [inline addons/account_facturx/data/facturx_templates.xml:238]: Is empty `rdf:about` really needed?

File: addons/account_facturx/data/facturx_templates.xml
Comment (smetl) [inline addons/account_facturx/data/facturx_templates.xml:267]: this line is too long

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [inline addons/account_facturx/models/ir_actions_report.py:29]: `fields.Date.context_today(self)` to handle timezone correctly

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [inline addons/account_facturx/models/ir_actions_report.py:26]: Empty `begin` attribute: is it needed?

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [inline addons/account_facturx/models/ir_actions_report.py:23]: You need to specify the params inside the docstring

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [inline addons/account_facturx/models/ir_actions_report.py:31]: The most efficient way is to concat the string only once like: 
```
metadata_content = self.env.ref('account_facturx.account_invoice_pdfa_3_facturx_metadata').render({...})
metadata = f'<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>{metadata_content}<?xpacket end="w"?>'
```

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [reply addons/account_facturx/models/ir_actions_report.py:26]: Are you sure `id` is a static value? It looks weird to me

File: addons/account_facturx/models/ir_actions_report.py
Comment (smetl) [inline addons/account_facturx/models/ir_actions_report.py:14]: always specify the keywords: here, `pdf_content=pdt_content`

File: odoo/tools/pdf.py
Comment (smetl) [inline odoo/tools/pdf.py:61]: when overriding an existing python class like this, the best practice is
```
def __init__(*args, **kwargs, pdf_content=None)
    super().__init__(*args, **kwargs)
```

File: odoo/tools/pdf.py
Comment (smetl) [inline odoo/tools/pdf.py:75]: IMO, you should follow the same behavior as the original class and make this part in `_add_facturx_attachments` and add the pdf-a parameter:
```
buffer = io.BytesIO(pdf_content)
reader = PdfFileReader(buffer)
writer = OdooPdfFileWriter()
writer.cloneDocumentFromReader(reader)
writer.enablePDFA(title="...", subject="...")
writer.write(...) # PDF-A should be handled when writing the final content
```

File: odoo/tools/pdf.py
Comment (smetl) [inline odoo/tools/pdf.py:343]: I don't really understand why you need all this stuff here: to be discussed

File: odoo/tools/pdf.py
Comment (smetl) [inline odoo/tools/pdf.py:121]: This `__init__` is dangerous and makes me think it might not be a good idea to handle the PDF-A inside the OdooPdfFileWriter: TO BE DISCUSSED

File: addons/account_facturx/data/facturx_templates.xml
Comment (vin-odoo) [reply addons/account_facturx/data/facturx_templates.xml:238]: It is written as such in the example given in the factur-x doc/zip

I didn't try without it but I kept it like in the example to be sure

File: addons/account_facturx/models/ir_actions_report.py
Comment (vin-odoo) [reply addons/account_facturx/models/ir_actions_report.py:26]: see: [adobe xmp doc](https://wwwimages2.adobe.com/content/dam/acom/en/devnet/xmp/pdfs/XMP%20SDK%20Release%20cc-2016-08/XMPSpecificationPart3.pdf) page 41
In an older version of the same document, it was explicitly said to use that id. I also compared with other valid factur-x pdf documents and they all had the exact same id.

Comment (vin-odoo) [reply addons/account_facturx/models/ir_actions_report.py:26]: As for the begin="", it is not empty, although it could be:  

> This required attribute indicates the beginning of a new packet. Its value is the Unicode zero-width nonbreaking space character U+FEFF, in the appropriate encoding (UTF-8, UTF-16, or UTF-32). It serves as a
byte-order marker, where the character is written in the natural order of the application (consistent with the byte
order of the XML data encoding).
For backward compatibility with earlier versions of the XMP packet specification, the value of this attribute can
be the empty string, indicating UTF-8 encoding.

**Review by smetl — APPROVED** [review]

**Comment by vin-odoo** [comment]
@odoo/rd-security Hi, could we get a review of this, please :) Thank you!

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:75]: Why are you not using a bytes literal and binary escapes? The last 4 characters are completely garbled.

The behaviour also seems debatable, what will happen to existing clients who don't care, is 1.7 and its "some more bytes" fully backwards compatible?

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:150]: Since you're not actually using the streaming capabilities of Popen, why not just use `subprocess.run`?

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:157]: So if pdfa is requested and we don't manage to generate it we just... ignore the entire thing?

And since we're purportedly looking up a binary beforehand where does the `FileNotFoundError` come from?

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:87]:
```suggestion
        Override the base class to delay the addition post ghostscript call, as well as properly add them to fit with
```

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:35]:
```suggestion
    return None
```

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:29]:
```suggestion
            run([command, '--version'], stdout=DEVNULL, stderr=DEVNULL, check=True)
```

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:78]: This requires a lot more explanations. Why md5, where does `_ID` come from, why does it insert the same data twice, ...

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
This seems to be backporting 0226901ec74368532b0518034845849a9b54df88 which... sure, why not (except it conflicts with the original because `addAttachment` has a completely different implementation), but that should be a separate and clearly explained commits.

I also don't see the reason for implementing the PDF/A stuff and calling ghoststript in core instead of just facturx, are there others who will need that? And if so, whom?

**Comment by smetl** [comment]
@xmo-odoo We will need PDF-A later in other module. Also, PDF-A is something that could be apply to any PDF. That's why it should be a core feature.

File: odoo/tools/pdf.py
Comment (vin-odoo) [reply odoo/tools/pdf.py]: From [Vera pdf validator](https://github.com/veraPDF/veraPDF-validation-profiles/wiki/PDFA-Parts-2-and-3-rules#rule-613-1)
> The first string is a permanent identifier based on the contents of the file at the time it was originally created and does not change when the file is incrementally updated. The second string is a changing identifier based on the file's contents at the time it was last updated. When a file is first written, both identifiers are set to the same value.

As for why md5: I based my code on the way Pypdf2 is creating this id. But it only does it when encrypting the pdf (as in the past, it was the only time where it was needed). But now with PDF/A, we need this information at all times.

Note: I have removed the manual creation of /ID to reuse the one from the incoming stream, so this code no longer exists.

File: odoo/tools/pdf.py
Comment (vin-odoo) [reply odoo/tools/pdf.py]: Indeed, thank you for pointing this out. Initially, we did not use a system parameter (it was asked by Anthony recently) so we decided to silently ignore the issues since, for most users, this was not a requirement.

But as we now enable it if we want it, it would make total sense to raise errors. I did it here but still waiting for confirmation from Antony that it's the right thing to do to raise errors here.

**Comment by vin-odoo** [comment]
@xmo-odoo Hi :wave: thank you for the review, the code has been updated.

As for the separate commit, I don't know :/ This is not really a backporting, we needed to reimplement the add attachment for this since the one from PyPDF2 does not set the "/AF" key in the pdf and it is required by PDF/A.
We also need to change the flow to add them after converting because Ghostscript tends to remove the attachments from the pdf from the test I have done, so it is safer like that to avoid potential issues.

_create_attachment_object is indeed something from that pr, but I would need to redo the same anyway so I took it to avoid conflicting on this method (although I'll be conflicting on everything else anyway)
As for the class name, it's again just to avoid conflicting with the more recent code.

So to me, it's more of a separate development that conflicts with the existing one in a later version than a backport.

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

* md5 checksums are specified by the PDF 1.7 spec
* `file_open` uses a static path to a related data file
* `sys.platform` is compeltely innocuous

File: odoo/tools/__init__.py
Comment (KangOl) [inline odoo/tools/__init__.py:7]: For compatibility, this import should stay.

File: addons/account_facturx/models/ir_actions_report.py
Comment (odony) [inline addons/account_facturx/models/ir_actions_report.py:31]: It is customary to use `tools.str2bool` rather than `literal_eval` for parsing boolean ICP values. That's nicer for users because they can use any of 0/1/yes/no/true/false/on/off case-insensitively with the expected result, rather than having to write Python literals exactly.

(it's also better because you don't want to evaluate arbitrary values, just yes/no, so you don't need the full power of literal_eval)

File: odoo/tools/pdf.py
Comment (KangOl) [inline odoo/tools/pdf.py:24]: When installed through homebrew, the binary is `gs`
```suggestion
        'darwin': ['gsos2', 'gs'],
```

File: odoo/tools/__init__.py
Comment (vin-odoo) [reply odoo/tools/__init__.py:7]: I see, changing the way it was imported is most likely a mistake I made. Should I remove the other one I added `from .pdf import *` and keep this one but put it below `from .translate import *`?
I moved it below because I am using the translations _() and it was included later than the pdf.
(And changed the way it's written to fit with the other ones there, but that was probably a bad idea)

File: addons/account_facturx/models/ir_actions_report.py
Comment (KangOl) [inline addons/account_facturx/models/ir_actions_report.py:23]:
```suggestion
                header = b'<?xpacket begin="﻿" id="W5M0MpCehiHzreSzNTczkc9d"?>'
                footer = b'<?xpacket end="w"?>'
```

File: odoo/tools/pdf.py
Comment (odony) [inline odoo/tools/pdf.py:118]: any reference to the magic bytes spec somewhere?

**Review by odony — COMMENTED** [review]
Thanks for the very nice work!

a few more suggestions to finalize it, and then we should be good to merge :)

File: addons/account_facturx/models/ir_actions_report.py
Comment (vin-odoo) [reply addons/account_facturx/models/ir_actions_report.py:18]: I don't think it is only relevant for PDF/A. As far as I know, this is a requirement for Factur-x that I added here as it was detected as an error by the zugferd validator.

File: odoo/tools/pdf.py
Comment (vin-odoo) [reply odoo/tools/pdf.py:55]: The OdooPdfFileWriter already exists in master (it was made to allow embedding multiple attachments in a single file, something PyPDF2 does not support). I reused that name for this reason, as it is reimplementing this functionality (but making the attachment PDF/A compliant too, even if not needed it doesn't hurt).

As for the f, it came from the method definition in master I reused it to keep it consistent, probably for "file", I can remove this.

File: odoo/tools/pdf.py
Comment (vin-odoo) [reply odoo/tools/pdf.py:132]: It does come from there indeed, the licensing for this particular file is found here: https://www.color.org/profiles2.xalter#license
(Which says the same as the one you have)

I added the copyright found in the file and text I found there "This profile is made available..." please let me know if this is right thank you.

File: addons/account_facturx/models/ir_actions_report.py
Comment (odony) [reply addons/account_facturx/models/ir_actions_report.py]: The ID of the template and its content clearly refer to PDF/A  (`http://www.aiim.org/pdfa/ns/id/` is the PDF/A namespace, etc.) Not sure it makes sense to have it in a facturx doc if the pdfa conversion isn't done. But it's fine for me to add it anyway, as long as we handle the case where it's missing gracefully :)

Comment (vin-odoo) [reply addons/account_facturx/models/ir_actions_report.py]: Indeed, my bad. I moved it to only add it with only PDF/A anyway as it does make sense, and the factur-x pdf file should always be PDF/A to be correct.

File: odoo/tools/pdf.py
Comment (odony) [reply odoo/tools/pdf.py:132]: that looks perfect, thanks! :+1:

File: addons/account_facturx/models/ir_actions_report.py
Comment (odony) [inline addons/account_facturx/models/ir_actions_report.py:31]: Not very important, but it is a best practice to avoid doing string formatting yourself when emitting logs, and instead leave that to the logger[1]: you simply pass the arguments after the log message. That makes the code a bit prettier, and among other things, it means the formatting cost does not need to be paid if the log message is silenced.

Besides, `%s` format always performs `str()` on its argument, so you don't need to do it at all.

```suggestion
                        _logger.exception("Error while converting to PDF/A: %s", e)
```

(The exception message will be printed twice because `logger.exception()` does it too, but that's fine, at least here it will be on the same line)

[1] https://docs.python.org/3/library/logging.html#logging.debug

Comment (vin-odoo) [reply addons/account_facturx/models/ir_actions_report.py]: Done thank you :)

File: odoo/tools/pdf.py
Comment (odony) [inline odoo/tools/pdf.py:13]: nitpicking: imports are normally sorted alphabetically inside each group, in order to make them easier to read/scan. Really minor, but as we're reordering them, might as well do it well ;)

we can keep fonttools separate as it gets special treatment

```suggestion
import io

from datetime import datetime
from hashlib import md5
from logging import getLogger
from PyPDF2 import PdfFileWriter, PdfFileReader
from PyPDF2.generic import DictionaryObject, NameObject, ArrayObject, DecodedStreamObject, NumberObject, createStringObject, ByteStringObject
from zlib import compress, decompress

try:
    from fontTools.ttLib import TTFont
except ImportError:
    TTFont = None
```

File: addons/account_facturx/models/ir_actions_report.py
Comment (odony) [inline addons/account_facturx/models/ir_actions_report.py:7]: order?

```suggestion
from io import BytesIO
from logging import getLogger
from PyPDF2 import PdfFileReader

from odoo import fields, models
from odoo import tools
from odoo.tools.pdf import OdooPdfFileWriter
```

**Comment by smetl** [comment]
@robodoo r+

**Comment by smetl** [comment]
@robodoo r+

**Pull request closed without merging**
